apps/frontend/static/frontend/api.js

```diff
@@ -0,0 +1,184 @@
+    }
+
+    function getEntityBuilder(cls) {
+        return function (id) {
+            return getEntity(cls, id);
+        }
+    }
+
+    function getEntityFetcher(cls, id) {
+        return function () {
+            return getEntity(cls, id);
+        }
+    }
+
+    function createEntityClass() {
+        return function (data) {
+            this.__data = data;
+            this.init();
+        }
+    }
+
+    var PagedList;
+
+    PagedList = function (cls) {
+        this.childCls = cls;
+        this.nextUrl = Config.apiUrl + cls.API_NAME + '/';
+        this.currentPage = [];
+        this.currentIndex = 0;
+
+    };
+    PagedList.prototype.next = function () {
+        var deferred = $.Deferred(),
+            cls = this.childCls;
+        if (this.currentIndex < this.currentPage.length) {
+            deferred.resolve(new cls(this.currentPage[this.currentIndex++]));
+        } else if (this.nextUrl) {
+            Api.call(this.nextUrl)
+                .done(function (r) {
+                    this.nextUrl = r.next;
+                    this.currentPage = r.results;
+                    this.currentIndex = 0;
+                    if (this.currentPage.length > 0) {
+                        deferred.resolve(new cls(this.currentPage[this.currentIndex++]));
+                    } else {
+                        deferred.reject();
+                    }
+                }.bind(this))
+                .fail(function (r) {
+                    deferred.reject.apply(deferred, arguments);
+                }.bind(this))
+        } else {
+            deferred.reject();
+        }
+        return deferred.promise();
+    };
+    PagedList.prototype.all = function (entryCb, endCb) {
+        var self = this;
+
+        function getNext() {
+            self.next()
+                .done(function (o) {
+                    entryCb && entryCb(o);
+                    getNext();
+                })
+                .fail(function () {
+                    endCb && endCb()
+                });
+        }
+
+        getNext();
+    };
+    PagedList.prototype.allAtOnce = function (allCb) {
+        var all = [];
+        this.all(function (r) {
+            all.push(r);
+        }, function () {
+            allCb && allCb(all);
+        });
+    };
+
+    function getPagedListBuilder(cls) {
+        return function () {
+            return new PagedList(cls);
+        }
+    }
+
+
+    CompareRequest = createEntityClass();
+    TestRun = createEntityClass();
+    Result = createEntityClass();
+    Task = createEntityClass();
+
+    CompareRequest.API_NAME = 'compare_requests';
+    TestRun.API_NAME = 'test_runs';
+    Result.API_NAME = 'results';
+    Task.API_NAME = 'tasks';
+
+    CompareRequest.RELATED_FIELDS = {
+        base_test_run: TestRun,
+        head_test_run: TestRun
+    };
+    TestRun.RELATED_ARRAYS = {
+        results: Result,
+        tasks: Task
+    };
+
+
+    var allClassess = [CompareRequest, TestRun, Result, Task];
```

Comment:
A minor one: all local variables are defined at the top of this function, worth adding this one there as well?